### ISSUE TITLE
poppler build fails

### ISSUE THREAD

**Issue by andreafioraldi** [HexHive/magma#31]
Hi, I'm trying to use magma with my own custom fuzzer, but the poppler build seems to fail:

```[ 98%] Building CXX object cpp/CMakeFiles/poppler-cpp.dir/poppler-rectangle.cpp.o
/magma/targets/poppler/repo/cpp/poppler-global.cpp:248:17: error: no matching function for call to 'iconv'
    size_t ir = iconv(ic, (ICONV_CONST char **)&me_data, &me_len_char, &str_data, &str_len_left);
                ^~~~~
/usr/include/iconv.h:42:15: note: candidate function not viable: no known conversion from 'const char **' to 'char **__restrict' for 2nd argument
extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
              ^
/magma/targets/poppler/repo/cpp/poppler-global.cpp:254:14: error: no matching function for call to 'iconv'
        ir = iconv(ic, (ICONV_CONST char **)&me_data, &me_len_char, &str_data, &str_len_left);
             ^~~~~
/usr/include/iconv.h:42:15: note: candidate function not viable: no known conversion from 'const char **' to 'char **__restrict' for 2nd argument
extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
              ^
/magma/targets/poppler/repo/cpp/poppler-global.cpp:302:17: error: no matching function for call to 'iconv'
    size_t ir = iconv(ic, (ICONV_CONST char **)&str_data, &str_len_char, &ret_data, &ret_len_left);
                ^~~~~
/usr/include/iconv.h:42:15: note: candidate function not viable: no known conversion from 'const char **' to 'char **__restrict' for 2nd argument
extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
              ^
/magma/targets/poppler/repo/cpp/poppler-global.cpp:308:14: error: no matching function for call to 'iconv'
        ir = iconv(ic, (ICONV_CONST char **)&str_data, &str_len_char, &ret_data, &ret_len_left);
             ^~~~~
/usr/include/iconv.h:42:15: note: candidate function not viable: no known conversion from 'const char **' to 'char **__restrict' for 2nd argument
extern size_t iconv (iconv_t __cd, char **__restrict __inbuf,
              ^
4 errors generated.
[100%] Linking CXX static library libpoppler-cpp.a
/usr/bin/ar: CMakeFiles/poppler-cpp.dir/poppler-global.cpp.o: No such file or directory
cpp/CMakeFiles/poppler-cpp.dir/build.make:406: recipe for target 'cpp/libpoppler-cpp.a' failed
make[3]: *** [cpp/libpoppler-cpp.a] Error 1
CMakeFiles/Makefile2:734: recipe for target 'cpp/CMakeFiles/poppler-cpp.dir/all' failed
make[2]: *** [cpp/CMakeFiles/poppler-cpp.dir/all] Error 2
make[1]: *** [cpp/CMakeFiles/poppler-cpp.dir/rule] Error 2
CMakeFiles/Makefile2:746: recipe for target 'cpp/CMakeFiles/poppler-cpp.dir/rule' failed
Makefile:383: recipe for target 'poppler-cpp' failed
make: *** [poppler-cpp] Error 2
```

The other targets are ok and this error should not be related to my fuzzer , I'm just using afl-clang-fast of AFL++ to instrument.

Maybe it is a cmake related error? https://gitlab.freedesktop.org/poppler/poppler/blob/master/cmake/modules/FindIconv.cmake#L26 here ICONV_CONST is set to const only after a test.

**Comment by hazimeh** [comment]
Does this happen inside the provided docker images?
Edit: Ah yes, just noticed the `/magma/targets/poppler/repo` prefix.

Would it be possible to docker-commit and docker-save the last container that fails the build and send it over?

**Comment by hazimeh** [comment]
Could you please post your CMake command output? In my docker image build:
```
-- The C compiler identification is Clang 9.0.0
-- The CXX compiler identification is Clang 9.0.0
-- Check for working C compiler: /magma/fuzzers/aflplusplus/repo/afl-clang-fast
-- Check for working C compiler: /magma/fuzzers/aflplusplus/repo/afl-clang-fast -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - failed
-- Check for working CXX compiler: /magma/fuzzers/aflplusplus/repo/afl-clang-fast++
-- Check for working CXX compiler: /magma/fuzzers/aflplusplus/repo/afl-clang-fast++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
...
-- Performing Test ICONV_SECOND_ARGUMENT_IS_CONST
-- Performing Test ICONV_SECOND_ARGUMENT_IS_CONST - Failed
-- Found Iconv: /usr/lib/x86_64-linux-gnu/libc.so
...
```
`ICONV_SECOND_ARGUMENT_IS_CONST` is False, and `ICONV_CONST` is thus empty.
If it were True, `ICONV_CONST` would be `const` and the compilation should not fail.

**Comment by andreafioraldi** [comment]
```-- The C compiler identification is Clang 10.0.1
-- The CXX compiler identification is Clang 10.0.1
-- Check for working C compiler: /magma/fuzzers/aflpp/repo/constr-clang
-- Check for working C compiler: /magma/fuzzers/aflpp/repo/constr-clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - failed
-- Check for working CXX compiler: /magma/fuzzers/aflpp/repo/constr-clang++
-- Check for working CXX compiler: /magma/fuzzers/aflpp/repo/constr-clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- Checking _FILE_OFFSET_BITS for large files
-- Checking _FILE_OFFSET_BITS for large files - not needed
^[[91mCMake Warning at CMakeLists.txt:78 (message):


   No test data found in $testdatadir.
   You will not be able to run 'make test' successfully.



   The test data is not included in the source packages
   and is also not part of the main git repository. Instead,
   you can checkout the test data from its own git
   repository with:



     git clone git://git.freedesktop.org/git/poppler/test



   You should checkout the test data as a sibling of your
   poppler source folder or specify the location of your
   checkout with -DTESTDATADIR=/path/to/checkoutdir/test.



^[[0m-- no fontconfig or win32 specific code
-- Found Freetype: /magma/targets/poppler/work/lib/libfreetype.a (found version "2.8.1")
-- Found JPEG: /usr/lib/x86_64-linux-gnu/libjpeg.so
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
-- Found PNG: /usr/lib/x86_64-linux-gnu/libpng.so (found version "1.6.34")
-- Found TIFF: /usr/lib/x86_64-linux-gnu/libtiff.so (found version "4.0.9")
-- Checking for module 'cairo>=1.10.0'
--   Found cairo, version 1.15.10
-- Found Cairo: /usr/lib/x86_64-linux-gnu/libcairo.so (Required is at least version "1.10.0")
-- Performing Test ICONV_SECOND_ARGUMENT_IS_CONST
-- Performing Test ICONV_SECOND_ARGUMENT_IS_CONST - Success
-- Found Iconv: /usr/lib/x86_64-linux-gnu/libc.so
-- The imported target "openjp2_static" references the file
   "/usr/lib/x86_64-linux-gnu/libopenjp2.a"
......
```

**Comment by andreafioraldi** [comment]
I'm using llvm 10, just take the preinstall script of afl++ lto and substitute 11 with 10

**Comment by andreafioraldi** [comment]
If it is not llvm 10, then it is my compiler wrapper (that it is just using the afl++ compiler pass in this configuration) but seems unlikely

**Comment by andreafioraldi** [comment]
> Would it be possible to docker-commit and docker-save the last container that fails the build and send it over?

It contains private work. I'll try to reproduce it using just afl++ if possible and send it to u then.

**Comment by andreafioraldi** [comment]
btw (unrelated), do u mind sharing the script that produces the graphs (or at least just the simple total unique crashes) in the sample report? I cannot find them in the source base and would be great to have those nice graphs

**Comment by hazimeh** [comment]
The first draft is on the dev branch:
https://github.com/HexHive/magma/blob/dev/tools/report_df/main.py

It generates only plots. The whole-report generation code is still in my local repo, as it needs some more refactoring and clean-up.

Edit: Refer to this for usage details:
https://github.com/HexHive/magma/issues/24#issuecomment-690797224

**Comment by hazimeh** [comment]
Closing this issue for lack of activity.
